File: iosApp/iosApp/Modules/KeyStore ✅/KeyStorePresenter.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3lib

enum KeyStorePresenterEvent {
    case didSelectKeyStoreItemtAt(idx: Int)
    case didSelectAccessory(idx: Int)
    case didSelectErrorActionAt(idx: Int)
    case didSelectButtonAt(idx: Int)
    case didChangeButtonsSheetMode(sheetMode: ButtonSheetViewModel.SheetMode)
}

protocol KeyStorePresenter: AnyObject {
    func present()
    func handle(_ event: KeyStorePresenterEvent)
}

final class DefaultKeyStorePresenter {
    private weak var view: KeyStoreView?
    private let wireframe: KeyStoreWireframe
    private let interactor: KeyStoreInteractor
    private let settingsService: SettingsService

    private var targetView: KeyStoreViewModel.TransitionTargetView = .none
    private var buttonsViewModel: ButtonSheetViewModel = .init(
        buttons: ButtonSheetViewModel.compactButtons(),
        sheetMode: .compact
    )

    init(
        view: KeyStoreView,
        wireframe: KeyStoreWireframe,
        interactor: KeyStoreInteractor,
        settingsService: SettingsService
    ) {
        self.view = view
        self.wireframe = wireframe
        self.interactor = interactor
        self.settingsService = settingsService
    }

    func updateView() {
        view?.update(with: viewModel())
    }
}

extension DefaultKeyStorePresenter: KeyStorePresenter {

    func present() {
        updateView()
    }

    func handle(_ event: KeyStorePresenterEvent) {
        switch event {
        case let .didSelectKeyStoreItemtAt(idx):
            handleDidSelectItem(at: idx)
        case let .didSelectAccessory(idx):
            handleDidSelectAccessory(at: idx)
        case let .didSelectErrorActionAt(idx: idx):
            handleDidSelectErrorAction(at: idx)
        case let .didSelectButtonAt(idx):
            handleButtonAction(at: idx)
        case let .didChangeButtonsSheetMode(mode):
            handleDidChangeButtonsState(mode)
        }
    }
}

private extension DefaultKeyStorePresenter {

    func handleDidSelectItem(at idx: Int) {
        let keyStoreItem = interactor.items[idx]
        interactor.selected = keyStoreItem
        view?.update(with: viewModel())
        wireframe.navigate(to: .networks)
    }

    func handleDidSelectAccessory(at idx: Int) {
        targetView = .keyStoreItemAt(idx: idx)
        view?.updateTargetView(targetView)
        let item = interactor.items[idx]
        wireframe.navigate(
            to: .keyStoreItem(
                item: item,
                handler: { [weak self] item in
                    self?.handleDidUpdateNewKeyStoreItem(item)
                },
                onDeleted: { [weak self] in
                    guard let self = self else { return }
                    if self.interactor.items.isEmpty {
                        self.wireframe.navigate(to: .hideNetworksAndDashboard)
                    }
                    self.present()
                }
            )
        )
    }

    func handleDidSelectErrorAction(at idx: Int) {
        view?.update(with: viewModel())
    }

    func handleButtonAction(at idx: Int) {
        targetView = .buttonAt(idx: idx)
        view?.updateTargetView(targetView)
        switch buttonsViewModel.buttons[idx].type {
        case .newMnemonic:
            wireframe.navigate(to: .newMnemonic { [weak self] keyStoreItem in
                 self?.handleDidCreateNewKeyStoreItem(keyStoreItem)
            })
        case .importMnemonic:
            wireframe.navigate(to: .importMnemonic { [weak self] item in
                 self?.handleDidCreateNewKeyStoreItem(item)
            })
        case .moreOption:
            handleDidChangeButtonsState(.expanded)
        case .connectHardwareWallet:
            wireframe.navigate(to: .connectHardwareWaller)
        case .importPrivateKey:
            wireframe.navigate(to: .importPrivateKey)
        case .createMultiSig:
            wireframe.navigate(to: .createMultisig)
        }
    }

    func handleDidChangeButtonsState(_ sheetMode: ButtonSheetViewModel.SheetMode) {
        guard buttonsViewModel.sheetMode != sheetMode else { return }
        buttonsViewModel = .init(
            buttons: sheetMode.isCompact()
                ? ButtonSheetViewModel.compactButtons()
                : ButtonSheetViewModel.expandedButtons(),
            sheetMode: sheetMode
        )
        updateView()
    }

    func handleDidCreateNewKeyStoreItem(_ keyStoreItem: KeyStoreItem) {
        interactor.selected = keyStoreItem
        if let idx = interactor.items.firstIndex(of: keyStoreItem) {
            targetView = .keyStoreItemAt(idx: idx)
        }
        updateView()
        if interactor.items.count == 1 {
            // HACK: This is non ideal, but since we dont have `viewDidAppear`
            // simply animate to dashboard after first wallet was created
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.2) { [weak self] in
                 self?.wireframe.navigate(to: .dashBoardOnboarding)
            }
        }
        targetView = .none
    }

    func handleDidUpdateNewKeyStoreItem(_ keyStoreItem: KeyStoreItem) {
        if let idx = interactor.items.firstIndex(of: keyStoreItem) {
            targetView = .keyStoreItemAt(idx: idx)
        }
        view?.update(with: viewModel())
    }
}

private extension DefaultKeyStorePresenter {

    func viewModel(_ state: KeyStoreViewModel.State = .loaded) -> KeyStoreViewModel {
        .init(
            isEmpty: interactor.items.isEmpty,
            state: state,
            items: interactor.items.map {
                KeyStoreViewModel.KeyStoreItem(
                    title: $0.name,
                    address: $0.addressFormatted
                )
            },
            selectedIdxs: [
                selectedIdxs()
            ].compactMap{ $0 },
            buttons: buttonsViewModel,
            targetView: targetView,
            transitionStyle: ServiceDirectory.transitionStyle
        )
    }
    
    func selectedIdxs() -> Int? {
        if let index = interactor.items.firstIndex(of: interactor.selected) { return index }
        // If we have items and the keystore item could not tell us which one is selected
        // we default to 0, this should not happen though!
        guard interactor.items.count > 0 else { return nil }
        return 0
    }

    func viewModel(from error: Error) -> KeyStoreViewModel {
        viewModel(
            .error(
                error: KeyStoreViewModel.Error(
                    title: "Error",
                    body: error.localizedDescription,
                    actions: [Localized("OK")]
                )
            )
        )
    }
}

private extension Array where Element == KeyStoreItem {
    
    func firstIndex(of keyStoreItem: KeyStoreItem?) -> Int? {
        firstIndex(where: { $0.uuid == keyStoreItem?.uuid })
    }
}

private extension KeyStoreItem {
    var addressFormatted: String? {
        guard let address = addresses[derivationPath] else { return nil }
        return Formatter.address.string(
            address,
            digits: 10,
            for: .ethereum()
        )
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/keyStore/KeyStorePresenter.kt
package com.sonsofcrypto.web3walletcore.modules.keyStore

import com.sonsofcrypto.web3lib.formatters.Formatters
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreItem
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3lib.utils.uiDispatcher
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStorePresenterEvent.*
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreViewModel.ButtonSheetViewModel.Button
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreViewModel.ButtonSheetViewModel.Button.Type.*
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreViewModel.ButtonSheetViewModel.SheetMode
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreViewModel.ButtonSheetViewModel.SheetMode.COMPACT
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreViewModel.ButtonSheetViewModel.SheetMode.EXPANDED
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreViewModel.Item
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreViewModel.State.Loaded
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreViewModel.TransitionTargetView.*
import com.sonsofcrypto.web3walletcore.modules.keyStore.KeyStoreWireframeDestination.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

sealed class KeyStorePresenterEvent {
    data class DidSelectKeyStoreItemtAt(val idx: Int): KeyStorePresenterEvent()
    data class DidSelectAccessory(val idx: Int): KeyStorePresenterEvent()
    data class DidSelectErrorActionAt(val idx: Int): KeyStorePresenterEvent()
    data class DidSelectButtonAt(val idx: Int): KeyStorePresenterEvent()
    data class DidChangeButtonsSheetMode(val mode: SheetMode): KeyStorePresenterEvent()
}

interface KeyStorePresenter {
    fun present()
    fun handle(event: KeyStorePresenterEvent)
}

class DefaultKeyStorePresenter(
    private val view: WeakRef<KeyStoreView>,
    private val wireframe: KeyStoreWireframe,
    private val interactor: KeyStoreInteractor,
): KeyStorePresenter {
    private var buttonsSheet: KeyStoreViewModel.ButtonSheetViewModel = buttonsCompacted()
    private var targetView: KeyStoreViewModel.TransitionTargetView = None
    private val uiScope = CoroutineScope(uiDispatcher)

    override fun present() { updateView() }

    override fun handle(event: KeyStorePresenterEvent) {
        when (event) {
            is DidSelectKeyStoreItemtAt -> handleDidSelectKeyStoreItem(event.idx)
            is DidSelectAccessory -> handleDidSelectAccessory(event.idx)
            is DidSelectErrorActionAt -> handleDidSelectErrorAction(event.idx)
            is DidSelectButtonAt -> handleButtonAction(event.idx)
            is DidChangeButtonsSheetMode -> handleDidChangeButtonsState(event.mode)
        }
    }

    private fun handleDidSelectKeyStoreItem(idx: Int) {
        val keyStoreItem = interactor.items[idx]
        interactor.selected = keyStoreItem
        updateView()
        wireframe.navigate(Networks)
    }

    private fun handleDidSelectAccessory(idx: Int) {
        targetView = KeyStoreItemAt(idx)
        view.get()?.updateTargetView(targetView)
        val item = interactor.items[idx]
        wireframe.navigate(
            EditKeyStoreItem(
                item,
                handler = {
                    val newIdx = interactor.items.indexOf(it)
                    if (newIdx != -1) { targetView = KeyStoreItemAt(newIdx) }
                    updateView()
                },
                onDeleted = {
                    if (interactor.items.isEmpty()) { wireframe.navigate(HideNetworksAndDashboard) }
                    present()
                }
            )
        )
    }

    private fun handleDidSelectErrorAction(idx: Int) { updateView() }

    private fun handleButtonAction(idx: Int) {
        targetView = ButtonAt(idx)
        view.get()?.updateTargetView(targetView)
        when (buttonsSheet.buttons[idx].type) {
            NEW_MNEMONIC -> wireframe.navigate(
                NewMnemonic { handleDidCreateNewKeyStoreItem(it) }
            )
            IMPORT_MNEMONIC -> wireframe.navigate(
                NewMnemonic { handleDidCreateNewKeyStoreItem(it) }
            )
            MORE_OPTION -> handleDidChangeButtonsState(EXPANDED)
            CONNECT_HARDWARE_WALLET -> wireframe.navigate(ConnectHardwareWallet)
            IMPORT_PRIVATE_KEY -> wireframe.navigate(ImportPrivateKey)
            CREATE_MULTI_SIG -> wireframe.navigate(CreateMultisig)
        }
    }

    private fun handleDidCreateNewKeyStoreItem(keyStoreItem: KeyStoreItem) {
        interactor.selected = keyStoreItem
        val idx = interactor.items.indexOf(keyStoreItem)
        if (idx != -1) { targetView = KeyStoreItemAt(idx) }
        updateView()
        targetView = None
        navigateToDashboardIfNeeded()
    }

    private fun navigateToDashboardIfNeeded() {
        if (interactor.items.count() != 1) return
        // HACK: This is non ideal, but since we dont have `viewDidAppear`
        // simply animate to dashboard after first wallet was created
        uiScope.launch {
            delay(1200.toLong())
            wireframe.navigate(DashboardOnboarding)
        }
    }

    private fun handleDidChangeButtonsState(mode: SheetMode) {
        if (buttonsSheet.mode == mode) return
        if (mode == COMPACT) {
            buttonsSheet = buttonsCompacted()
        } else if (mode == EXPANDED) {
            buttonsSheet = buttonsExpanded()
        }
        updateView()
    }

    private fun updateView(state: KeyStoreViewModel.State = Loaded) {
        view.get()?.update(viewModel(state))
    }

    private fun viewModel(state: KeyStoreViewModel.State): KeyStoreViewModel = KeyStoreViewModel(
        interactor.items.isEmpty(),
        state,
        interactor.items.map { Item(it.name, it.addressFormatted) },
        listOf(selectedIdxs()).mapNotNull { it },
        buttonsSheet,
        targetView
    )

    private fun selectedIdxs(): Int? {
        val index = interactor.items.indexOfFirst { it == interactor.selected }
        if (index != -1) return index
        // If we have items and the keystore item could not tell us which one is selected
        // we default to 0, this should not happen though!
        return if (interactor.items.isEmpty()) null else 0
    }

    private fun buttonsCompacted(): KeyStoreViewModel.ButtonSheetViewModel =
        KeyStoreViewModel.ButtonSheetViewModel(
            listOf(
                Button(Localized("keyStore.newMnemonic"), NEW_MNEMONIC),
                Button(Localized("keyStore.importMnemonic"), IMPORT_MNEMONIC),
                Button(Localized("keyStore.moreOption"), MORE_OPTION),
                Button(Localized("keyStore.importPrivateKey"), IMPORT_PRIVATE_KEY),
                Button(Localized("keyStore.createMultiSig"), CREATE_MULTI_SIG)
            ),
            COMPACT
        )

    private fun buttonsExpanded(): KeyStoreViewModel.ButtonSheetViewModel =
        KeyStoreViewModel.ButtonSheetViewModel(
            listOf(
                Button(Localized("keyStore.newMnemonic"), NEW_MNEMONIC),
                Button(Localized("keyStore.importMnemonic"), IMPORT_MNEMONIC),
                Button(Localized("keyStore.connectHardwareWallet"), CONNECT_HARDWARE_WALLET),
                Button(Localized("keyStore.importPrivateKey"), IMPORT_PRIVATE_KEY),
                Button(Localized("keyStore.createMultiSig"), CREATE_MULTI_SIG)
            ),
            EXPANDED
        )
}

private val KeyStoreItem.addressFormatted: String? get() {
    val address = addresses[derivationPath] ?: return null
    // TODO: Review here when supporting other networks
    return Formatters.networkAddress.format(address, 10, Network.ethereum())
}
